mm2src/mm2_p2p/src/behaviours/peers_exchange.rs

```diff
@@ -0,0 +1,394 @@
+        let mut rng = rand::thread_rng();
+        let peer_ids = self
+            .known_peers
+            .clone()
+            .into_iter()
+            .filter(|peer| !self.request_response.addresses_of_peer(peer).is_empty())
+            .collect::<Vec<_>>();
+
+        let peer_ids = peer_ids.choose_multiple(&mut rng, num);
+        for peer_id in peer_ids {
+            let addresses = self.request_response.addresses_of_peer(peer_id).into_iter().collect();
+            result.insert((*peer_id).into(), addresses);
+        }
+        result
+    }
+
+    fn forget_peer(&mut self, peer: &PeerId) {
+        self.known_peers.retain(|known_peer| known_peer != peer);
+        self.forget_peer_addresses(peer);
+    }
+
+    fn forget_peer_addresses(&mut self, peer: &PeerId) {
+        for address in self.request_response.addresses_of_peer(peer) {
+            if !self.is_reserved_peer(peer) {
+                self.request_response.remove_address(peer, &address);
+            }
+        }
+    }
+
+    pub fn add_peer_addresses_to_known_peers(&mut self, peer: &PeerId, addresses: PeerAddresses) {
+        for address in addresses.iter() {
+            if !self.validate_global_multiaddr(address) {
+                warn!("Attempt adding a not valid address of the peer '{}': {}", peer, address);
+                return;
+            }
+        }
+        if !self.known_peers.contains(peer) && !addresses.is_empty() {
+            self.known_peers.push(*peer);
+        }
+        let already_known = self.request_response.addresses_of_peer(peer);
+        for address in addresses {
+            if !already_known.contains(&address) {
+                self.request_response.add_address(peer, address);
+            }
+        }
+    }
+
+    pub fn add_peer_addresses_to_reserved_peers(&mut self, peer: &PeerId, addresses: PeerAddresses) {
+        for address in addresses.iter() {
+            if !self.validate_global_multiaddr(address) {
+                return;
+            }
+        }
+
+        if !self.reserved_peers.contains(peer) && !addresses.is_empty() {
+            self.reserved_peers.push(*peer);
+        }
+
+        let already_reserved = self.request_response.addresses_of_peer(peer);
+        for address in addresses {
+            if !already_reserved.contains(&address) {
+                self.request_response.add_address(peer, address);
+            }
+        }
+    }
+
+    fn maintain_known_peers(&mut self) {
+        if self.known_peers.len() > MAX_PEERS {
+            let mut rng = rand::thread_rng();
+            let to_remove_num = self.known_peers.len() - MAX_PEERS;
+            self.known_peers.shuffle(&mut rng);
+            let removed_peers: Vec<_> = self.known_peers.drain(..to_remove_num).collect();
+            for peer in removed_peers {
+                self.forget_peer_addresses(&peer);
+            }
+        }
+        self.request_known_peers_from_random_peer();
+    }
```

Comment:
We will need a better mechanism for removing peers when we have too many. We should make an effort to stayed peered with good peers. This random selection introduces potential eclipse attack issues.

Reply:
c.c. @onur-ozkan